velox/functions/prestosql/aggregates/ChecksumAggregate.cpp

```diff
@@ -0,0 +1,228 @@
+      } else {
+        vector->set(i, *value<int64_t>(group));
+      }
+    }
+  }
+
+  void addRawInput(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    auto hashes = getHashBuffer(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
+        computeHashForNull(groups[row]);
+      } else {
+        computeHash(groups[row], rawHashes[row]);
+      }
+    });
+  }
+
+  void addIntermediateResults(
+      char** groups,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+
+    rows.applyToSelected(
+        [&](vector_size_t i) { *value<int64_t>(groups[i]) += rawValues[i]; });
+  }
+
+  void addSingleGroupRawInput(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    const auto& arg = args[0];
+    auto hasher = getPrestoHasher(arg->type());
+    auto hashes = getHashBuffer(rows.end(), arg->pool());
+    hasher->hash(arg, rows, hashes);
+    auto rawHashes = hashes->as<int64_t>();
+
+    rows.applyToSelected([&](vector_size_t row) {
+      if (arg->isNullAt(row)) {
+        computeHashForNull(group);
+      } else {
+        computeHash(group, rawHashes[row]);
+      }
+    });
+  }
+
+  void addSingleGroupIntermediateResults(
+      char* group,
+      const SelectivityVector& rows,
+      const std::vector<VectorPtr>& args,
+      bool /**/) override {
+    clearNull(group);
+
+    auto vector = args[0]->asUnchecked<FlatVector<int64_t>>();
+    auto rawValues = vector->rawValues();
+    int64_t result = 0;
+    if (vector->mayHaveNulls()) {
+      rows.applyToSelected([&](vector_size_t i) {
+        if (!vector->isNullAt(i)) {
+          result += rawValues[i];
+        }
+      });
+    } else {
+      rows.applyToSelected([&](vector_size_t i) { result += rawValues[i]; });
+    }
+
+    *value<int64_t>(group) += result;
+  }
+
+ private:
+  FOLLY_ALWAYS_INLINE void computeHash(char* group, const int64_t hash) {
+    clearNull(group);
+    auto result = *value<int64_t>(group);
+    result += hash * kPrime64;
+    *value<int64_t>(group) = result;
+  }
+
+  FOLLY_ALWAYS_INLINE void computeHashForNull(char* group) {
+    clearNull(group);
+    *value<int64_t>(group) += kPrime64;
+  }
+
+  FOLLY_ALWAYS_INLINE PrestoHasher* getPrestoHasher(TypePtr typePtr) {
+    if (prestoHasher_ == nullptr) {
+      prestoHasher_ = std::make_unique<PrestoHasher>(std::move(typePtr));
+    }
+    return prestoHasher_.get();
+  }
+
+  FOLLY_ALWAYS_INLINE BufferPtr& getHashBuffer(
+      vector_size_t size,
```

Comment:
The callers seem to be passing size as the number of entries, but hashes_->size() is the number of bytes, e.g. 8 * number-of-entries. Hence, the check below seems wrong.